### DISCUSSION TITLE
Chartbrew v3.0

### DISCUSSION THREAD

**Discussion by razvanilin** [chartbrew/chartbrew#177] [General]
> Feel free to bookmark this as I will update the discussion as we progress towards the v3 release

**Hey everyone!**

V2 was a major shift over v1 and due to the extended usage in both managed and self-hosted versions, we gathered a lot of data that we can use to re-work and improve the product moving forwards. The changes we are going to work on are so big that it warrants a major version change to `v3.0`

The main changes that will come to v3 are going to change the way Chartbrew is used. These will affect the way Connections, Datasets, and Projects work.

Currently, each **Project** comes with its own **Connections** and each **Chart** comes with its own **Datasets** at the project level. This essentially means that each project is completely separate from the other with no way of transferring information. This is great for the flexibility it gives and the amount of resources available (e.g. lots of different connections spread across different projects) but the real usage shows that users are not using it this way - it's more of an inconvenience.

### The major v3 change

To answer the inconvenience, we sat down and carefully designed a better experience for new and existing users. The focus will now shift toward Connections and Datasets.

First of all, architecture-wise, the connections, and datasets will both shift to the team level. This means that each project will be able to use the same connections and datasets. If a connection or dataset updates, it will reflect across all projects and charts. Don't want changes to propagate over different projects? You can solve this by creating new connections and datasets the same way it works now in v2.

**The data will be migrated automatically so it will not affect existing dashboards**

Next, the general flow will change:

- v2: Create project -> create connection -> create chart -> create dataset
- v3: Create connection -> create dataset -> create project -> create chart (from existing datasets)

**A better experience for teams:**

This change also takes into account teams with different skills. The connections can be created by people with the credentials to data sources, the datasets can be created with queries by technical&data people, and charts can be created by sales&marketing.

I will update this discussion if there are any plan changes, details, and screenshots to share. Also, feel free to [join our discord](https://discord.gg/KwGEbFk) where you can see more discussions about v3 and Chartbrew in general.

### Checklist

> Some of these changes might be added to Chartbrew before the formal release of v3

- [ ] Team-based Connections
- [ ] Team-based Datasets
- [ ] New Team UI
- [ ] New connection manager UI
- [ ] New dataset manager UI
- [ ] Updated chart creation UI
- [ ] Drag&drop dashboard editor to arrange and resize charts
- [ ] More chart sizes (height + width)
- [ ] New dashboard elements - title and description text blocks

> ETA end of November 2023

**Comment by hdarch** [comment]
Please, add tagged autocomplete (https://www.npmjs.com/package/react-tag-autocomplete) instead of plain input textbox for filters as well as an option to show added filters on Dashboard Repots page

**Reply by razvanilin** [reply]
Thanks for the suggestion @hdarch Will look into it 👍

**Comment by razvanilin** [comment]
Due to the changes in the overall architecture, the current team roles will have to change a bit. This will unfortunately delay the release, but will improve the team management a lot.

Currently, most roles (except the owner) are project-level, but this has to change because the connections and datasets are moving to the team-level.

The new roles will be:

**Team-level**: Team owner (full project access), Team admin (full project access), Project-only member (specific project access)
**Project-level**: Project admin, Project viewer

This will require a migration of current roles. These are the default switches:

Owner -> Team owner
Admin -> Project-only member, Project admin (to all current ones they can access)
Editor -> Project-only member, Project admin (to all current ones they can access)
Member -> Project-only member, Project viewer (to all current ones they can access)

**Please note that only the owner will be moved to the team-level roles. The rest of the admins will have to be added manually if needed.**

Here's a visual diagram to get a better idea:

![image](https://github.com/chartbrew/chartbrew/assets/5342321/0d72236f-2fcf-4331-9a6b-f01cd7ca25cc)
